README.md
grunt-sauce-driver
==================

[![Build Status](https://travis-ci.org/burnnat/grunt-sauce-driver.png)](https://travis-ci.org/burnnat/grunt-sauce-driver) [![Dependency Status](https://gemnasium.com/burnnat/grunt-sauce-driver.png)](https://gemnasium.com/burnnat/grunt-sauce-driver) [![NPM version](https://badge.fury.io/js/grunt-sauce-driver.png)](http://badge.fury.io/js/grunt-sauce-driver)

[![Test Status](https://saucelabs.com/browser-matrix/grunt-sauce-driver.svg)](https://saucelabs.com/u/grunt-sauce-driver)

A [Grunt](http://gruntjs.com/) task for running [WebDriver scripts](https://github.com/admc/wd) in the [Sauce Labs](https://saucelabs.com/) cloud.

In addition to supporting custom scripts, standard scripts are provided for running javascript unit tests in [Jasmine](http://pivotal.github.io/jasmine/) and [Siesta Lite](http://www.bryntum.com/products/siesta/).

Inspired by the [grunt-saucelabs](https://github.com/axemclion/grunt-saucelabs) plugin, this project aims to provide greater flexibility and support for custom WebDriver scripts, as well as the option to run tests locally.

Local Tests
-----------

Local tests are currently supported for Google Chrome and Internet Explorer. To run tests locally, ensure you have downloaded the appropriate [ChromeDriver](https://sites.google.com/a/chromium.org/chromedriver/home) or [InternetExplorerDriver](http://code.google.com/p/selenium/wiki/InternetExplorerDriver) server for your platform and made it available on your `PATH` environment variable.

Usage
-----

### Example ###

```javascript
var saucedriver = require('grunt-sauce-driver');

grunt.initConfig({
	saucedriver: {
		options: {
			username: 'saucelabs-user-name',
			key: 'saucelabs-key',
			tunneled: true,
			tags: ['example', 'sample'],
			browsers: [{
				browserName: 'opera'
			}]
		},
		
		jasmine: {
			url: 'http://example.com/test/jasmine.html',
			script: saucedriver.jasmine
		},
		
		siesta: {
			url: 'http://example.com/test/siesta.html',
			script: saucedriver.siesta
		},
		
		custom: {
			url: 'http://example.com/test/custom.html',
			script: function(browser, options) {
				return (
					browser
					.waitForElementByClassName('some-class')
					.elementByXPath('//input[type=button]')
					.then(function(el) {
						return browser.moveTo(el);
					})
					.click()
					.log('Test complete')
				);
			}
		}
	}
});
```

### Options ###

#### Basic Options ####

* __url__ : String _Required_

  The URL of the page to be tested.

* __script__ : Function _Required_

  The WebDriver script to run, either a custom function or one provided by `grunt-sauce-driver`.
  The function should accept the following parameters:

  * __browser__: The WebDriver [browser](https://github.com/admc/wd#browser-initialization) object (using the [promises](https://github.com/admc/wd#q-promises--chaining) API).
  * __options__: The Grunt options object.

* __browsers__ : Object[] _Required_

  An array of objects representing the [browsers](https://saucelabs.com/docs/platforms) on which this test should run.

* __concurrency__ : Number _Optional_

  Number of concurrent browsers to test against. When running on Sauce Labs, check your account info for your maximum allowed concurrency. Defaults to `1`.

* __slow__ : Boolean/Number _Optional_

  Adds a delay between each WebDriver action, to make tests easier to watch and debug. Can be specified either as a boolean flag, or a number of milliseconds to delay. If set to `true`, a default value of 500 milliseconds will be used. Defaults to `false`.

* __logging__ : Boolean _Optional_

  If enabled, logs information about the running script to the console. Defaults to `true`.

* __testTimeout__ : Number _Optional_

  Number of milliseconds to wait for javascript tests to complete on each page before timeout and failing the test. Default to `180000`.

* __testInterval__ : Number _Optional_

  Number of milliseconds between each poll to see if a javascript test is complete. Defaults to `2000`.

* __testReadyTimeout__ : Number _Optional_

  Number of milliseconds to wait for the test page to load. Defaults to `10000`.

* __mirrorTestFailure__ : Boolean _Optional_

  When enabled, the grunt task will fail whenever any of the javascript tests fail. Otherwise, the task will fail only on WebDriver error (page load failure, test timeout, etc). Defaults to `true`.

* __ignoreFailure__ : Boolean _Optional_

  If enabled, specifies that the grunt task should always pass regardless of WebDriver errors or javascript test failures. (The task may still fail under certain catastrophic errors, however.) Defaults to `false`.

#### SauceLabs Options ####

* __username__ : String _Optional_

  The username that will be used to connect to Sauce Labs. Defaults to `process.env.SAUCE_USERNAME`.

* __key__ : String _Optional_

  The access key to provide when connecting Sauce Labs. For security purposes, don't hard-code this value in your Gruntfile. Instead, load it via local config file or environment variable. Defaults to `process.env.SAUCE_ACCESS_KEY`.

* __tunneled__ : Boolean _Optional_

  If enabled, uses a [Sauce Connect](https://saucelabs.com/docs/connect) tunnel when running tests on Sauce Labs. Defaults to `true`.

* __tunnelTimeout__ : Number _Optional_

  The connection timeout to use when initializing Sauce Connect, in seconds. Defaults to `120`.

* __testname__ : String _Optional_

  The name of the test, to be displayed on the Sauce Labs dashboard. Defaults to `""`.

* __tags__ : String[] _Optional_

  An array of tags for this test, used when filtering the Sauce Labs dashboard. Defaults to `[]`.

* __build__ : String _Optional_

  The build identifier to be associated with this test. Defaults to `null`.

#### Local Options ####

* __local__ : Boolean _Optional_

  If enabled, will run the current test on the local machine rather than Sauce Labs. Defaults to `false`.

* __autoclose__ : Boolean _Optional_

  When running local tests, determines whether the browser under test will be automatically closed after the test completes. Defaults to `true`.

* __driverPorts__ : Object _Optional_

  A map of browser names to port numbers, for use when communicating with the local WebDriver server. Defaults to `{ "chrome": 9515, "internet explorer": 5555 }`.

License
-------

This project is published under the terms of the [MIT License](http://opensource.org/licenses/MIT).

Copyright &copy; 2013 by [Nat Burns](https://github.com/burnnat).


CHANGELOG.md
# 0.1.1 / 2013-11-23

 * [BUGFIX] Fix error logging non-element action arguments.

# 0.1.0 / 2013-10-19

 * [FEATURE] Add support for Internet Explorer when running locally.
 * [ENHANCEMENT] Update to use chained promises API from WebDriver 0.2.0.
 * [ENHANCEMENT] Add ability to specify delay time when running in "slow" mode.
 * [BUGFIX] Fix Javascript errors when running Siesta tests in older browsers.

# 0.0.3 / 2013-10-10

 * [FEATURE] Add `mirrorTestFailure` and `ignoreFailure` options for finer control of test pass/fail status.
 * [ENHANCEMENT] Automatically retry failed tests in Siesta, to work around SauceLabs browser issues with running Siesta.
 * [BUGFIX] Ignore any unsupported browsers when running locally.

# 0.0.2 / 2013-10-07

 * [ENHANCEMENT] Terminate task immediately with error message if any required options are missing.
 * [BUGFIX] Prevent from hanging when no custom data is supplied by the test script.

# 0.0.1 / 2013-10-06

 * Initial release.

test/siesta/basics/index.js
var Harness = Siesta.Harness.Browser

Harness.configure({
    title     : 'Awesome Test Suite',
    autoCheckGlobals : true,

    // no files to preload
    preload : [
    ]
})


Harness.start(
    // plain strings - urls
    '010_basic_assertions.t.js',
    '020_async_code.t.js',
    
    // and object with `url` property and its own "preload" option
    {
        url : '030_global_variables.t.js',
        preload : [
            // Jquery CDN
            'http://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js'
        ]
    },
    
    '040_todo_tests.t.js',

    '050_wait_for.t.js',

    '060_bdd.t.js'
)



test/siesta/basics/030_global_variables.t.js
StartTest({
    preload : [
        // Jquery CDN
        'http://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js'
    ]
}, function (t) {
    t.diag('Testing simple for loop...');

    // If you find a global leak, you can use this trick to catch it (uncomment the for loop below first, and run in a modern browser)
//    Object.defineProperty(window, 'i', {
//        set : function() { debugger; }
//    })

    // Since this test is preloading jQuery, we should let Siesta know what to 'expect'
    t.expectGlobals('$', 'jQuery');

//    // try to uncomment this line to see the report about unexpected global - "i"
//    // please note: this feature is not supported in IE <= 8
//    for (i = 0; i < 10; i++) {}
    
    t.pass('For loop executed ok, no exceptions.');
    
    $('body').html('JQuery was here');

    t.like(document.body.innerHTML, 'JQuery was here', 'Found correct text in DOM');
});


Gruntfile.js
var saucedriver = require('./drivers.js');

module.exports = function(grunt) {
	grunt.initConfig({
		connect: {
			server: {
				options: {
					port: 9999
				}
			}
		},
		
		saucedriver: {
			options: {
				browsers: [
//					{
//						browserName: 'safari',
//						platform: 'OS X 10.8'
//					},
					{
						browserName: 'opera',
						platform: 'Linux'
					},
					{
						browserName: 'chrome',
						platform: 'Windows 7'
					},
					{
						browserName: 'firefox',
						version: '25',
						platform: 'Windows 7'
					},
					{
						browserName: 'internet explorer',
						platform: 'Windows 8',
						version: '10'
					},
					{
						browserName: 'internet explorer',
						platform: 'Windows 7',
						version: '9'
					},
					{
						browserName: 'internet explorer',
						platform: 'Windows XP',
						version: '8'
					}
				],
				
				build: process.env.TRAVIS_JOB_ID,
				local: grunt.option('local'),
				concurrency: 3
			},
			
			jasmine: {
				options: {
					url: 'http://localhost:9999/test/jasmine/examples/spec-runner/spec-runner.html',
					script: saucedriver.jasmine,
					testname: 'jasmine tests'
				}
			},
			
			siesta: {
				options: {
					url: 'http://localhost:9999/test/siesta/basics/index.html',
					script: saucedriver.siesta,
					testname: 'siesta tests'
				}
			}
		}
	});
	
	grunt.loadTasks('tasks');
	
	// Loading dependencies
	for (var key in grunt.file.readJSON("package.json").devDependencies) {
		if (key !== "grunt" && key !== "grunt-cli" && key.indexOf("grunt") === 0) {
			grunt.loadNpmTasks(key);
		}
	}
	
	grunt.registerTask("test", ["connect", "saucedriver"]);
};


reporters/siesta.js
Siesta.REPORTER = true;

Siesta.Harness.meta.extend({
	methods: {
		generateJSONReport: function() {
			return {
				durationSec: (this.endDate - this.startDate) / 1000,
				passed: this.allPassed(),
				suites: Ext.Array.map(
					this.descriptors,
					function(descriptor) {
						var suite = this.getSuiteResults(
							this.testsByURL[descriptor.url].getResults().toJSON()
						);
						
						if (suite.totalCount) {
							return Ext.apply(
								Ext.copyTo(
									{},
									suite,
									[
										'description',
										'durationSec',
										'passed'
									]
								),
								{
									specs: [
										Ext.apply(suite, { description: '[all]' })
									]
								}
							);
						}
						else {
							return suite;
						}
					},
					this
				)
			};
		},
		
		getSuiteResults: function(results) {
			var output = {
				description: results.name || results.url.replace(/^.*\//, ''),
				durationSec: (results.endDate - results.startDate) / 1000,
				passed: results.passed
			};
			
			var suites = [];
			var specs = [];
			
			var pass = 0;
			var fail = 0;
			var total = 0;
			
			Ext.Array.forEach(
				results.assertions,
				function(child) {
					if (child.type === "Siesta.Result.SubTest") {
						(child.bddSpecType === 'describe'
								? suites
								: specs)
							.push(
								this.getSuiteResults(child)
							);
					}
					else {
						if (child.passed)
							pass++;
						else {
							fail++;
						}
						
						total++;
					}
				},
				this
			);
			
			if (suites.length || specs.length) {
				return Ext.apply(output, {
					suites: suites,
					specs: specs
				});
			}
			else {
				return Ext.apply(output, {
					passedCount: pass,
					failedCount: fail,
					totalCount: total
				});
			}
		}
	}
});

tasks/lib/sauce-runner.js
var _ = require('lodash');
var async = require('async');

var EventEmitter = require('events').EventEmitter;
var webdriver = require('wd');
var SauceLabs = require('saucelabs');
var SauceTunnel = require('sauce-tunnel');

var locals = require('./local-browsers');

module.exports = {
	
	createTunnel: function(grunt, options, callback) {
		var _ = grunt.util._;
		
		if (this.tunnel) {
			callback(this.tunnel);
			return;
		}
		
		var tunnel = this.tunnel = new SauceTunnel(
			options.username,
			options.key,
			options.identifier,
			true,
			options.tunnelTimeout
		);
		
		var methods = ['write', 'writeln', 'error', 'ok', 'debug'];
		methods.forEach(function (method) {
			tunnel.on('log:' + method, function (text) {
				grunt.log[method](text);
			});
			
			tunnel.on('verbose:' + method, function (text) {
				grunt.verbose[method](text);
			});
		});
		
		grunt.log.writeln("=> Connecting to Saucelabs ...");
		
		tunnel.start(function(created) {
			if (!created) {
				grunt.log.error('Failed to create SauceConnect tunnel.');
				return callback(null);
			}
			
			grunt.log.ok("Connected to Saucelabs.");
			
			/*
			 * Okay, so this is pretty hacky... but we can't perform async cleanup
			 * using process.on('exit'), so we monkey-patch the grunt exit method
			 * to allow us to perform our cleanup and complete the exit afterward.
			 */
			grunt.util.exit = _.wrap(
				grunt.util.exit,
				function(original, code) {
					var scope = this;
					var args = _.last(arguments, 1);
					
					tunnel.stop(function() {
						grunt.log.writeln('Tunnel connection closed.');
						original.call(scope, args);
					});
				}
			);
			
			process.on(
				'exit',
				function() {
					tunnel.stop(function() {
						grunt.log.writeln('Tunnel connection closed.');
					});
				}
			);
			
			callback(tunnel);
		});
	},
	
	drive: function(grunt, driverConfig, options, callback) {
		var _ = grunt.util._;
		
		var logError = function(message, error) {
			message = message + ': ' + (_.isObject(error) ? error.message : error);
			
			if (error.data) {
				var data = JSON.parse(error.data);
				
				if (data.value && data.value.message) {
					message = message + ' [ ' + data.value.message + ' ]';
				}
			}
			
			grunt.log.error(message);
		};
		
		var queue = async.queue(
			function(browserConfig, rootCallback) {
				var browser = webdriver.promiseChainRemote(driverConfig);
				
				var browserCallback = browserConfig.callback
					? function(err) {
						browserConfig.callback(err, browser, rootCallback);
					}
					: rootCallback;
				
				if (options.logging) {
					browser.on('status', function(info){
						grunt.log.writeln('\x1b[36m%s\x1b[0m', info);
					});
					
					browser.on('log', function(message){
						grunt.log.writeln(' > \x1b[36mLOG\x1b[0m: %s', message);
					});
					
					browser.on('command', function(method, path){
						grunt.log.writeln(' > \x1b[33m%s\x1b[0m: %s', method, path);
					});
					
					browser.on('promise', function(context, method, args, status) {
						var context = context === browser
							? '[]'
							: '[element ' + context.value + ']';
						
						args = _.clone(args);
						
						if (_.isFunction(_.last(args))) {
							args.pop();
						}
						
						args = args.map(function(arg) {
							if (_.isFunction(arg) && arg instanceof browser._Element) {
								return '[element ' + arg.toString() + ']';
							}
							else {
								return JSON.stringify(arg);
							}
						});
						
						grunt.log.verbose.writeln(
							' --> %s %s %s ( %s )',
							status,
							context,
							method,
							args.join(', ')
						);
					});
				}
				
				var scriptCallback = function(err) {
					if (browser.saucePassed === false) {
						var message = 'Javascript tests failed.';
						
						if (options.mirrorTestFailure) {
							err = message;
						}
						else {
							grunt.log.error(message);
						}
					}
					
					if (err) {
						logError('Script ended with error', err);
					}
					else {
						grunt.log.ok('Script run completed.');
					}
					
					browserCallback(err);
				};
				
				var delay = options.slow;
				
				if (delay) {
					var ignore = /^chain$|^toString$|^_/;
					
					_.each(
						_.functions(browser),
						function(name) {
							if (ignore.test(name) || EventEmitter.prototype[name]) {
								return;
							}
							
							var original = browser[name];
							
							browser[name] = function() {
								return original.apply(this, arguments).delay(delay);
							};
						}
					);
				}
				
				browser.init(browserConfig)
					.fail(function(err) {
						logError('Unable to initialize browser', err);
						browserCallback(err);
					})
					.get(options.url)
					.then(function() {
						return options.script(browser, options);
					})
					.fin(function() {
						if (options.autoclose) {
							return browser.quit();
						}
					})
					.done(
						function() {
							scriptCallback();
						},
						function(err) {
							scriptCallback(err);
						}
					);
			},
			options.concurrency
		);
		
		var failures;
		
		queue.drain = function() {
			if (failures) {
				grunt.log.error('One or more tests failed.');
			}
			
			callback(failures);
		};
		
		options.browsers.forEach(function(browserConfig) {
			var browserId = [browserConfig.browserName, browserConfig.version, browserConfig.platform].join(':');
			
			browserConfig.name = options.testname;
			browserConfig.tags = options.tags;
			browserConfig.build = options.build;
			
			var tunnelId = options['tunnel-identifier'];
			
			if (tunnelId) {
				browserConfig['tunnel-identifier'] = tunnelId;
			}
			
			browserConfig.callback = browserConfig.callback || options.callback;
			
			queue.push(
				browserConfig,
				function(err) {
					if (err) {
						failures = failures || {};
						failures[browserId] = err;
					}
				}
			);
		});
	},
	
	run: function(grunt, options, done) {
		if (!options.local) {
			if (options.tunneled) {
				var me = this;
				
				this.createTunnel(
					grunt,
					options,
					function(tunnel) {
						if (!tunnel) {
							done(false);
						}
						else {
							options['tunnel-identifier'] = tunnel.identifier;
							
							me.runRemote(
								grunt,
								options,
								function(err) {
									done(options.ignoreFailure || !err);
								}
							);
						}
					}
				);
			}
			else {
				this.runRemote(grunt, options, done);
			}
		}
		else {
			var matches = [];
			
			options.browsers.forEach(function(browser) {
				var name = browser.browserName;
				var platform = browser.platform;
				
				if (locals.hasBrowser(browser)) {
					grunt.log.verbose.writeln('Adding browser: ' + name);
					
					matches.push({
						browserName: name
					});
				}
				else {
					grunt.log.verbose.writeln("Skipping unsupported browser: " + name + " on " + platform)
				}
			});
			
			matches = _.unique(matches, 'browserName');
			
			if (matches.length < 1) {
				grunt.log.error("No supported local browsers found");
				return done(false);
			}
			
			var me = this;
			var failed = false;
			
			async.each(
				matches,
				function(browser, callback) {
					var localOptions = _.defaults({ browsers: [browser] }, options);
					
					var name = browser.browserName;
					var port = localOptions.driverPorts[name];
					
					grunt.log.writeln("=> Starting local WebDriver for browser '%s' on port %d ...", name, port);
					
					var scriptErr;
					
					var driver = locals.startDriver(browser, port);
					
					driver.on('error', function(error) {
						if (error.code === 'ENOENT') {
							driver = null;
							grunt.log.error('Unable to locate local WebDriver.');
							failed = true;
							callback();
						}
						else {
							throw error;
						}
					});
					
					driver.stderr.on('data', function(data) {
						grunt.log.verbose.writeln(data);
					});
					
					driver.on('close', function(code) {
						if (!driver) {
							return;
						}
						
						if (code) {
							grunt.log.error('Local WebDriver terminated with error: ' + code);
						}
						else {
							grunt.log.writeln('Local WebDriver terminated.');
						}
						
						failed = failed || (!localOptions.ignoreFailure && scriptErr);
						callback();
					});
					
					me.drive(
						grunt,
						{
							host: '127.0.0.1',
							port: port,
							path: '/'
						},
						localOptions,
						function(err) {
							scriptErr = err;
							locals.stopDriver(driver);
						}
					);
				},
				function() {
					done(!failed);
				}
			);
		}
	},
	
	runRemote: function(grunt, options, done) {
		var sauce = new SauceLabs({
			username: options.username,
			password: options.key
		});
		
		// Update SauceLabs job status as browser run completes.
		options.callback = function(browserError, browser, next) {
			var job = browser.sessionID;
			
			sauce.updateJob(
				job,
				{
					passed: !browserError && browser.saucePassed !== false
				},
				function(err) {
					if (err) {
						err = err.errors;
						grunt.log.error('Error updating SauceLabs status for job ' + job + ': ' + err);
					}
					else {
						grunt.log.verbose.writeln('Updated SauceLabs status for job: ' + job);
					}
					
					if (browser.sauceData) {
						sauce.updateJob(
							job,
							{
								'custom-data': browser.sauceData
							},
							function(warn) {
								if (warn) {
									grunt.log.warn('Error updating SauceLabs data for job ' + job + ': ' + warn.errors);
								}
								else {
									grunt.log.verbose.writeln('Updated SauceLabs data for job: ' + job);
								}
								
								next(browserError || err);
							}
						);
					}
					else {
						next(browserError || err);
					}
				}
			);
		};
		
		this.drive(
			grunt,
			{
				host: 'ondemand.saucelabs.com',
				port: 80,
				username: options.username,
				accessKey: options.key
			},
			options,
			done
		);
	}
};


tasks/lib/local-browsers.js
var fs = require('fs');
var path = require('path');
var spawn = require('child_process').spawn;

var os = {
	win32: /^(windows|vista|xp)/i,
	linux: /^linux/i,
	darwin: /^(mac|os x)/i
};

var platformTest = os[process.platform];

module.exports = {
	drivers: {
		'chrome': 'chromedriver',
		'internet explorer': 'IEDriverServer'
	},
	
	hasBrowser: function(browser) {
		return platformTest.test(browser.platform) && this.commandOnPath(this.drivers[browser.browserName]);
	},
	
	commandOnPath: function(command) {
		if (!command) {
			return false;
		}
		
		var delimiter = path.delimiter
		
		var extensions = process.env.PATHEXT
			? process.env.PATHEXT.split(delimiter)
			: [''];
		
		return (
			process.env.PATH.split(delimiter)
				.some(function(envpath) {
					return extensions.some(function(extension) {
						return fs.existsSync(
							path.join(envpath, command + extension)
						);
					});
				})
		);
	},
	
	startDriver: function(browser, port) {
		return spawn(
			this.drivers[browser.browserName],
			['--port=' + port]
		);
	},
	
	stopDriver: function(driver) {
		driver.kill();
	}
};

tasks/saucedriver.js
var runner = require('./lib/sauce-runner.js');

module.exports = function(grunt) {
	grunt.registerMultiTask(
		'saucedriver',
		'Run Selenium tests on SauceLabs via NodeJS WebDriver bindings',
		function() {
			var options = this.options({
				// Required options
				url: null,
				script: null,
				browsers: [],
				
				// Selenium options
				concurrency: 1,
				
				slow: false,
				logging: true,
				
				testTimeout: 180000,
				testInterval: 2000,
				testReadyTimeout: 10000,
				
				mirrorTestFailure: true,
				ignoreFailure: false,
				
				// SauceLabs options
				username: process.env.SAUCE_USERNAME,
				key: process.env.SAUCE_ACCESS_KEY,
				identifier: Math.floor((new Date()).getTime() / 1000 - 1230768000).toString(),
				tunneled: true,
				tunnelTimeout: 120,
				
				testname: "",
				tags: [],
				build: null,
				
				// Local options
				local: false,
				autoclose: true,
				driverPorts: {
					'chrome': 9515,
					'internet explorer': 5555
				}
			});
			
			if (!options.url) {
				grunt.log.error('No test URL specified.');
				return false;
			}
			
			if (!options.script) {
				grunt.log.error('No Selenium script specified.');
				return false;
			}
			
			if (!options.browsers || options.browsers.length < 1) {
				grunt.log.error('No target browsers specified.');
				return false;
			}
			
			if (options.slow === true) {
				options.slow = 500;
			}
			
			runner.run(grunt, options, this.async());
		}
	);
};